### ISSUE TITLE
model.half() does not seem to result in lower GPU memory usage

### ISSUE THREAD

**Issue by kingpalethe** [guillaume-be/rust-bert#202]
Thanks for this great library. Picking up from here: https://github.com/guillaume-be/rust-bert/issues/183#issuecomment-939260071

... I am testing the usage of `model.half()`. My guess is that the majority of users of `half()` will be using it (as I am) in order to fit larger models onto smaller amounts of available GPU memory.

I have tested the code provided in this issue comment verbatim, except that instead of `GPT_NEO_125M` I used `GPT_NEO_2_7B`, in order to test with a big model.

I ran the code while simultaneously monitoring GPU memory usage with `nvidia-smi`, and found that this code, at peak processing, used about 22.6GB of video RAM on my `rtx 3090`.

I tested it *with* `.half()`

```
    let mut model = TextGenerationModel::new(generate_config)?;
    model.half();
    model.set_device(Device::cuda_if_available());
```

and *without* `.half()`
```
  let model = TextGenerationModel::new(generate_config)?;
```

In *both* cases, the model used  about 22.6GB of video RAM... therefore, I think that the `.half()` method is either not actually reducing the size of the model, or something else is going wrong.

-------
(minor note)
I also had the idea that actually this `half()` method needs to be called directly on the `VarStore`, instead of the model, so also tried forking this project and adding a line here: https://github.com/guillaume-be/rust-bert/blob/b444780c18d617298a5f9f0497d2f67f69576004/src/gpt_neo/gpt_neo_model.rs#L679 ... I tried adding `var_store.half();`  ... but that was also not successful.

---

thanks!

**Comment by guillaume-be** [comment]
Hello @kingpalethe ,

I ran some local experiments and I am unfortunately unable to reproduce the issue you are facing. Using the script provided in the previous issue, I tried putting breakpoints after loading the model (for the parameters memory use) and after running a forward pass (for the parameters and activations), and monitor the output of `nvidia-smi`. 

Using a `GPT_NEO_1_3B` model variant, I get the following results, indicating memory savings in line with the half-precision conversion expectations:

| Precision | Memory use (model parameters) | Memory use (parameters + activations) |
|-----------|--------------------------------------|--------------------------------------------|
| Single     | 5296 MB                                         | 6238 MB                                                 |
| Half        | 2920 MB                                         | 3754 MB                                                 |

I ran these experiments on a RTX 2070 with 8GB, with drivers 471.41 and CUDA11.3

**Comment by kingpalethe** [comment]
Thanks for this and apologies for my error. I thought I had reproduced your sample code directly, but I had not, and I want to point out the error I made in case it is useful for anyone else in the future. 
![sequence-of-events](https://user-images.githubusercontent.com/11775831/143494579-f7e01a81-f802-4533-be17-5ab6c7dcfb22.png)
In your sample code, above, the model is (1) Loaded on the CPU, then (2) set to `half()` and then (3) moved to the GPU. 

My error was that I was initially loading it to the GPU in (1) instead with ` device: Device::cuda_if_available(),` so I think I was actually loading it twice onto the GPU, resulting the huge memory usage.
